File: bench/suite/src/st/fasta/go/fasta.go
// fasta: single-threaded, same algorithm as the Benchmarks Game description.
// Linear congruential generator, cumulative probabilities with linear search.
package main

import (
	"bufio"
	"os"
	"strconv"
)

const (
	im   = 139968
	ia   = 3877
	ic   = 29573
	line = 60
)

const alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
	"GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
	"CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
	"ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
	"GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
	"AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
	"AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

type acid struct {
	c byte
	p float64
}

var seed uint32 = 42

func random(max float64) float64 {
	seed = (seed*ia + ic) % im
	return max * float64(seed) / im
}

func repeatFasta(out *bufio.Writer, header string, src string, n int) {
	out.WriteString(header)
	l := len(src)
	buf := []byte(src + src[:line])
	pos := 0
	for left := n; left > 0; {
		m := line
		if left < line {
			m = left
		}
		out.Write(buf[pos : pos+m])
		out.WriteByte('\n')
		pos += m
		if pos >= l {
			pos -= l
		}
		left -= m
	}
}

func randomFasta(out *bufio.Writer, header string, table []acid, n int) {
	out.WriteString(header)
	chars := make([]byte, len(table))
	cumul := make([]float64, len(table))
	acc := 0.0
	for i, a := range table {
		acc += a.p
		chars[i] = a.c
		cumul[i] = acc
	}
	last := len(table) - 1
	var buf [line + 1]byte
	for left := n; left > 0; {
		m := line
		if left < line {
			m = left
		}
		for k := 0; k < m; k++ {
			r := random(1.0)
			i := 0
			for i < last && r >= cumul[i] {
				i++
			}
			buf[k] = chars[i]
		}
		buf[m] = '\n'
		out.Write(buf[:m+1])
		left -= m
	}
}

func main() {
	n := 1000
	if len(os.Args) > 1 {
		n, _ = strconv.Atoi(os.Args[1])
	}
	iub := []acid{
		{'a', 0.27}, {'c', 0.12}, {'g', 0.12}, {'t', 0.27},
		{'B', 0.02}, {'D', 0.02}, {'H', 0.02}, {'K', 0.02},
		{'M', 0.02}, {'N', 0.02}, {'R', 0.02}, {'S', 0.02},
		{'V', 0.02}, {'W', 0.02}, {'Y', 0.02},
	}
	homo := []acid{
		{'a', 0.3029549426680},
		{'c', 0.1979883004921},
		{'g', 0.1975473066391},
		{'t', 0.3015094502008},
	}
	out := bufio.NewWriterSize(os.Stdout, 1<<16)
	defer out.Flush()
	repeatFasta(out, ">ONE Homo sapiens alu\n", alu, 2*n)
	randomFasta(out, ">TWO IUB ambiguity codes\n", iub, 3*n)
	randomFasta(out, ">THREE Homo sapiens frequency\n", homo, 5*n)
}

File: bench/suite/src/st/fasta/rust/fasta.rs
// fasta: single-threaded, same algorithm as the Benchmarks Game description.
// Linear congruential generator, cumulative probabilities with linear search.
use std::io::{self, Write};

const IM: u32 = 139968;
const IA: u32 = 3877;
const IC: u32 = 29573;
const LINE: usize = 60;

const ALU: &[u8] = b"GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG\
GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA\
CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT\
ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA\
GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG\
AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC\
AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

struct Rng(u32);

impl Rng {
    fn next(&mut self, max: f64) -> f64 {
        self.0 = (self.0 * IA + IC) % IM;
        max * self.0 as f64 / IM as f64
    }
}

fn repeat_fasta<W: Write>(out: &mut W, header: &str, src: &[u8], n: usize) -> io::Result<()> {
    out.write_all(header.as_bytes())?;
    let len = src.len();
    // Doubled source lets every line be copied as one contiguous slice.
    let mut buf = Vec::with_capacity(len + LINE);
    buf.extend_from_slice(src);
    buf.extend_from_slice(&src[..LINE]);
    let mut pos = 0;
    let mut left = n;
    while left > 0 {
        let line = if left < LINE { left } else { LINE };
        out.write_all(&buf[pos..pos + line])?;
        out.write_all(b"\n")?;
        pos += line;
        if pos >= len {
            pos -= len;
        }
        left -= line;
    }
    Ok(())
}

fn random_fasta<W: Write>(
    out: &mut W,
    header: &str,
    table: &[(u8, f64)],
    n: usize,
    rng: &mut Rng,
) -> io::Result<()> {
    out.write_all(header.as_bytes())?;
    let mut chars = Vec::with_capacity(table.len());
    let mut cumul = Vec::with_capacity(table.len());
    let mut acc = 0.0;
    for &(c, p) in table {
        acc += p;
        chars.push(c);
        cumul.push(acc);
    }
    let last = table.len() - 1;
    let mut line = [0u8; LINE + 1];
    let mut left = n;
    while left > 0 {
        let m = if left < LINE { left } else { LINE };
        for slot in line.iter_mut().take(m) {
            let r = rng.next(1.0);
            let mut i = 0;
            while i < last && r >= cumul[i] {
                i += 1;
            }
            *slot = chars[i];
        }
        line[m] = b'\n';
        out.write_all(&line[..=m])?;
        left -= m;
    }
    Ok(())
}

fn main() -> io::Result<()> {
    let n: usize = std::env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(1000);
    let iub: [(u8, f64); 15] = [
        (b'a', 0.27), (b'c', 0.12), (b'g', 0.12), (b't', 0.27),
        (b'B', 0.02), (b'D', 0.02), (b'H', 0.02), (b'K', 0.02),
        (b'M', 0.02), (b'N', 0.02), (b'R', 0.02), (b'S', 0.02),
        (b'V', 0.02), (b'W', 0.02), (b'Y', 0.02),
    ];
    let homo: [(u8, f64); 4] = [
        (b'a', 0.3029549426680),
        (b'c', 0.1979883004921),
        (b'g', 0.1975473066391),
        (b't', 0.3015094502008),
    ];
    let stdout = io::stdout();
    let mut out = io::BufWriter::with_capacity(1 << 16, stdout.lock());
    let mut rng = Rng(42);
    repeat_fasta(&mut out, ">ONE Homo sapiens alu\n", ALU, n * 2)?;
    random_fasta(&mut out, ">TWO IUB ambiguity codes\n", &iub, n * 3, &mut rng)?;
    random_fasta(&mut out, ">THREE Homo sapiens frequency\n", &homo, n * 5, &mut rng)?;
    out.flush()
}
